app/workers/maintenance/poll_pagerduty.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+require 'pagerduty/maintenance_client'
+
+module PagerDuty
+  class PollMaintenanceWindows
+    include Sidekiq::Worker
+    include SentryLogging
+
+    MESSAGE_INDICATOR = 'USER_MESSAGE:'
+
+    def parse_user_message(raw_description)
+      raw_description.partition(MESSAGE_INDICATOR)[2].strip
+    end
+
+    def perform
+      client = PagerDuty::MaintenanceClient.new
+      pd_windows = client.get_all
+
+      # Add or update-in-place any open PagerDuty maintenance windows
+      pd_windows.each do |pd_win|
+        pd_win[:description] = parse_user_message(pd_win[:description])
+        window = MaintenanceWindow.find_or_initialize_by(pagerduty_id: pd_win[:pagerduty_id],
+                                                         external_service: pd_win[:external_service])
+        window.update(pd_win)
+      end
+
+      # Delete any existing records that are not present in the PagerDuty API results
+      # These indicate deleted maintenance windows so we want to stop reporting them
+      open_ids = pd_windows.map { |x| x[:pagerduty_id] }
+      MaintenanceWindow.end_after(Time.zone.now).each do |api_win|
```

Comment:
Does this maintain a record on our end for already ended maintenance windows that were deleted in PagerDuty?

Reply:
They'll remain in the database for potential extraction via a rake task or similar, but not be exposed via any API.

Maintenance windows that are deleted/canceled at any point before their original end time are removed from the database so there wouldn't be a record of those.